search/src/Components/upload/uploader_category_mapper.js

```diff
@@ -0,0 +1,112 @@
+import React from 'react';
+import { withStyles } from '@material-ui/core/styles';
+import FormControl from '@material-ui/core/FormControl';
+import Select from '@material-ui/core/Select';
+import InputLabel from '@material-ui/core/InputLabel';
+import MenuItem from '@material-ui/core/MenuItem';
+import TextField from '@material-ui/core/TextField';
+import Button from '@material-ui/core/Button';
+import cloneDeep from 'lodash/cloneDeep';
+
+
+const useStyles = (theme) => ({
+    summary: {
+        marginLeft: '2.5em'
+    },
+    row: {
+        display: 'flex',
+        alignItems: 'baseline'
+    },
+    formControl: {
+        margin: theme.spacing(1),
+        minWidth: 120,
+    },
+    text_field: {
+        marginRight: '1em'
+    },
+    applyButton: {
+        float: 'right'
+    }
+});
+
+class CategoryMapper extends React.Component {
+    
+    constructor(props) {
+        super(props);
+        this.state = {
+            categories: this.props.categories
+        }
+        this.modifyCategories = this.modifyCategories.bind(this);
+        this.changeCategory = this.changeCategory.bind(this);
+        this.changeSciName = this.changeSciName.bind(this);
+    }
+
+    modifyCategories(index, field, value) {
+        const newCategories = cloneDeep(this.state.categories);
+        newCategories[index][field] = value;
+        this.setState({categories: newCategories});
+    }
+
+    changeCategory(e, index) {
+        this.modifyCategories(index, "role", e.target.value);
+        this.props.handleCategoryStatus(false);
+        this.props.handleErrorMessage("init");
+    }
+
+    changeSciName(e, index) {
+        this.modifyCategories(index, "scientific_name", e.target.value);
+        this.props.handleCategoryStatus(false);
+        this.props.handleErrorMessage("init");
+    }
+
+    render() {
+        const { classes } = this.props;
+        return (
+            <React.Fragment>
+                <p className={classes.summary}>{this.state.categories.length} {this.state.categories.length != 1 ? 'categories' : 'category'} need mapping to weedcoco categories</p>
```

Comment:
This idea doesn't look good to me. Firstly, how do we define a complete row? For those user already uploaded with both role and sci_name or for those they have clicked the apply button for once? then they cannot change their mind after applying it?
The case is that there won't be say 100 categories to fill in for each upload. I reckon just let user have an overview what they will upload. Every time they change the form, it will be required to click the Apply button again to sync it and enable Next button. Otherwise it's messy to just manage the state of categories.
The best way is to give a checkbox of only showing not-yet-applied categories, and that's a bit over engineering if there will be just a few.

Reply:
Above I defined nComplete... just the way you'd defined it in the onclick handler: https://github.com/Sydney-Informatics-Hub/Weed-AI/pull/443/files#r629820294

I should be clear: I don't think we need to make this minor UI change to colour the rows when they're incomplete. It was just to communicate how I would see a more polished version of this.

A more polished version would also have a different apply/next interaction, but what you've got here is fine.

Reply:
I think reducing the number of errors as the user fills rows in is good, or at worst harmless.